File: Difficulty: Easy/Square root of a number/square-root-of-a-number.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// Function to find square root
// x: element to find square root
class Solution {
  public:
    long long int floorSqrt(long long int n) {
         if (n == 0 || n == 1) return n;
        
        long long int start = 1, end = n, ans;
        
        while (start <= end) {
            long long int mid = start + (end - start) / 2;
            
            if (mid <= n / mid) {
                ans = mid;
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        long long n;
        cin >> n;
        Solution obj;
        cout << obj.floorSqrt(n) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/Square root of a number/square-root-of-a-number.java
//{ Driver Code Starts
import java.util.Scanner;

class SquartRoot {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0) {
            long a = sc.nextInt();
            Solution obj = new Solution();
            System.out.println(obj.floorSqrt(a));
            t--;
        }
    }
}
// } Driver Code Ends


/*You are required to complete
this function*/

// Function to find square root
// x: element to find square root
class Solution {
     long floorSqrt(long n) {
         if(n==1)return n;
        long ans=-1;
        long hi = n/2;
        long lo = 1;
        while(hi>=lo)
        {
            long mid = ((hi+lo)/2);
            long sqr = mid*mid;
            if(sqr==n){
                ans =(long) mid;
                return ans;
            }else if(sqr>n)hi=(long)(mid-1);
            else{
                ans=(long)mid;
                lo = (long)mid+1;
            }
        }
        return ans;
    }
}
